### PR TITLE
Implement a label based discovery for Kafka Dev Service

### PR DESCRIPTION
**Pull request by cescoffier**
Instead to have to configure the kafka.bootstrap.url, it looks for containers having a specific label and computes the bootstrap server.
Discovery can be disabled in the configuration, and it is only used in dev mode (disabled for test).

@stuartwdouglas we discussed this briefly on the mailing list. It's a draft but works. 

**TODO:**

- [x] update the documentation
- [x] if we decide to go this way, move the label name to a shared class (so every dev service can use it)
- [x] extend configuration to allows configuring the label or value

### CODE REVIEW

**Comment by sberyozkin** [comment]
@cescoffier This looks great and will solve the problem I'm seeing with Dev Services for Keycloak - where a 2nd container is starting if I press Continuous Testing `r`.

**Comment by Ladicek** [comment]
I went through several dev services implementations in the last few days and I think we really need a common class that helps implementing a dev service based on a single container. That could also handle this label-based sharing.

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java [hunk 1]
```diff
@@ -35,6 +39,8 @@
     private static final Logger log = Logger.getLogger(DevServicesKafkaProcessor.class);
     private static final int KAFKA_PORT = 9092;
     private static final String KAFKA_BOOTSTRAP_SERVERS = "kafka.bootstrap.servers";
+    private static final String DEV_SERVICE_LABEL = "quarkus-dev-service";
```
Comment (stuartwdouglas) [inline extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java:42]: I wonder if this should be configurable, in case you want some isolated clusters.

Comment (cescoffier) [reply extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java:42]: We can either make the label configurable or the value. 
Right now, it's "kafka", but we can allow the user to configure it to something like "my-own-kafka-broker".

Comment (cescoffier) [reply extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java:42]: (I've added a todo)

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java [hunk 2]
```diff
@@ -253,6 +288,7 @@ private RedPandaKafkaContainer(DockerImageName dockerImageName, int fixedExposed
             this.port = fixedExposedPort;
             withNetwork(Network.SHARED);
             withExposedPorts(KAFKA_PORT);
+            withLabel(DEV_SERVICE_LABEL, DEV_SERVICE_NAME);
```
Comment (Ladicek) [inline extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java:291]: Should be added only in dev mode, and not in test mode?

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaDevServicesBuildTimeConfig.java [hunk 3]
```diff
@@ -33,4 +33,13 @@
...
     public String imageName;
 
+    /**
+     * Enables or disables the label-based service discovery allowing to look for a running Kafka before starting one.
+     * The discovery uses container labels to find a broker that has been launched by Dev Services for Kafka already.
+     *
+     * Discovery is only used in dev mode.
+     */
+    @ConfigItem(defaultValue = "true")
+    public boolean enableDiscovery;
```
Comment (Ladicek) [inline extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaDevServicesBuildTimeConfig.java:43]: Personally, I'd call this `shared`:

```java
/**
 * Enables sharing of a Kafka broker between multiple Quarkus dev mode applications.
 * If enabled, and another Quarkus dev mode application already launched Dev Services
 * for Kafka, the running broker will automatically be discovered, based on container labels,
 * and this application will be configured to use it.
 * <p>
 * Sharing is only used in dev mode, not for tests.
 */  
@ConfigItem(defaultValue = "true")
boolean shared;
```

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java (hunk 1 above)
Comment (cescoffier) [reply extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java:42]: Fixed.

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaDevServicesBuildTimeConfig.java (hunk 3 above)
Comment (cescoffier) [reply extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaDevServicesBuildTimeConfig.java]: Fixed.

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java (hunk 2 above)
Comment (cescoffier) [reply extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java]: Fixed.

**Comment by cescoffier** [comment]
@stuartwdouglas What do you want to do about this one? Should be move forward? 
(I will do the rebase later)

**Comment by stuartwdouglas** [comment]
I think it looks good.

**Comment by cescoffier** [comment]
Ok, I will rebase , polish a bit and ask for reviews.

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java [hunk 4]
```diff
@@ -101,10 +107,12 @@ public void run() {
         cfg = configuration;
 
         if (bootstrapServers != null) {
-            log.infof(
-                    "Dev Services for Kafka started. Start applications that need to use the same Kafka broker "
-                            + "using -Dkafka.bootstrap.servers=%s",
-                    bootstrapServers.getBootstrapServers());
+            if (kafkaBroker.isOwner()) {
+                log.infof(
+                        "Dev Services for Kafka started. Other Quarkus applications in dev mode will find the broker automatically. To make sure to use the same Kafka broker, "
```
Comment (stuartwdouglas) [inline extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java:112]: I think this is a bit confusing, as you say it works automatically then say you need the property. I would say something like 'For Quarkus applications in production mode you can connect to this by starting your application with ...'

File: core/deployment/src/main/java/io/quarkus/deployment/DevServices.java [hunk 5]
```diff
@@ -0,0 +1,11 @@
+package io.quarkus.deployment;
+
+public interface DevServices {
+
+    /**
+     * Label to add to shared Dev Services running in containers.
+     * This allows other applications to discover the running service and use it instead of starting a new instance.
+     */
+    String DEV_SERVICE_LABEL = "quarkus-dev-service";
```
Comment (stuartwdouglas) [inline core/deployment/src/main/java/io/quarkus/deployment/DevServices.java:9]: I wonder if this should be specific to the devservices type, i.e. quarkus-dev-services-kafka. If everything uses the same name and we enabled shared support for other things you may end up with kafka connecting to postgresql if a user misconfigures something.

**Review by stuartwdouglas — CHANGES_REQUESTED** [review]
Looks good, just some minor comments.

File: core/deployment/src/main/java/io/quarkus/deployment/DevServices.java (hunk 5 above)
Comment (cescoffier) [reply core/deployment/src/main/java/io/quarkus/deployment/DevServices.java:9]: Labels are key/value pairs, so while the key is the same, the value identifies the service. The lookup search for a container with the matching key/value and then verify that the expected port is mapper (exposed). 
We can have a specific one for each service, but we should make sure they start with the same prefix so we can identify the containers started by dev services.

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java (hunk 4 above)
Comment (cescoffier) [reply extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java:112]: Good point!

**Comment by cescoffier** [comment]
@stuartwdouglas I updated the code to use a specific label for Kafka.

**Review by stuartwdouglas — APPROVED** [review]

File: docs/src/main/asciidoc/kafka-dev-services.adoc [hunk 6]
```diff
@@ -25,6 +25,20 @@ Dev Services for Kafka relies on Docker to start the broker.
 If your environment does not support Docker, you will need to start the broker manually, or connect to an already running broker.
 You can configure the broker address using `kafka.bootstrap.servers`.
 
+== Shared broker
+
+Most of the time you need to share the broker between applications.
+Dev Services for Kafka implements a _service discovery_ mechanism to your multiple Quarkus applications running in _dev_ mode to share a single broker.
```
Comment (Ladicek) [inline docs/src/main/asciidoc/kafka-dev-services.adoc:31]:
```suggestion
Dev Services for Kafka implements a _service discovery_ mechanism for your multiple Quarkus applications running in _dev_ mode to share a single broker.
```

File: docs/src/main/asciidoc/kafka-dev-services.adoc [hunk 7]
```diff
@@ -25,6 +25,20 @@ Dev Services for Kafka relies on Docker to start the broker.
...
+
+NOTE: Dev Services for Kafka starts the container with the `quarkus-dev-service` label which is used to identify the container.
```
Comment (Ladicek) [inline docs/src/main/asciidoc/kafka-dev-services.adoc:33]:
```suggestion
NOTE: Dev Services for Kafka starts the container with the `quarkus-dev-service-kafka` label which is used to identify the container.
```

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java [hunk 8]
```diff
@@ -145,14 +181,31 @@ private KafkaBroker startKafka(KafkaDevServiceCfg config) {
...
+            log.warn(
+                    "Docker isn't working, please configure the Kafka bootstrap servers property (kafka.bootstrap.servers).");
             return null;
         }
 
+        if (config.shared && launchMode.getLaunchMode() == LaunchMode.DEVELOPMENT) {
+            // Detect if there is a broker already started using container labels.
+            Container container = lookup(config.serviceName);
+            if (container != null) {
+                log.infof("Dev Services for Kafka container found: %s (%s)", container.getId(),
```
Comment (Ladicek) [inline extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java:193]: Move this log message into the `if` below? Because it seems that at this point, it is _not_ clear that the shared broker will be used. Nice side effect is that there's gonna be only one log message, which is more than enough :-)

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaDevServicesBuildTimeConfig.java [hunk 9]
```diff
@@ -33,4 +33,31 @@
     @ConfigItem(defaultValue = "vectorized/redpanda:v21.5.5")
     public String imageName;
 
+    /**
+     * Indicates if the Kafka broker managed by Quarkus Dev Services is shared.
+     * When shared, it looks for running containers using a the label-based service discovery.
+     * If a matching container is found, it uses it, and so avoid to start a second one.
```
Comment (Ladicek) [inline extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaDevServicesBuildTimeConfig.java:39]: Too many its. "It uses it" is ... bad. I recall I suggested a vastly different text and it took me a while to write it. But the term "Dev Services for Kafka" is just too long to be used as a subject in a sentence, unless you take extra care.

Let me try again:

```java
/**
 * Indicates if the Kafka broker managed by Quarkus Dev Services is shared.
 * When shared, Quarkus looks for running containers using label-based service discovery.
 * If a matching container is found, it is used, and so a second one is not started.
 * Otherwise, Dev Services for Kafka starts a new container.
 * <p>
 * The discovery uses the {@code quarkus-dev-service-kafka} label. The value is configured
 * using the {@code service-name} property.
 * <p>
 * Container sharing is only used in dev mode.
 */
```

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaDevServicesBuildTimeConfig.java [hunk 10]
```diff
@@ -33,4 +33,31 @@
...
+     * The discovery uses the {@code quarkus-dev-service} label. The value is configured using the {@code service-name}
+     * property.
+     * <p>
+     * Sharing containers is only used in dev mode.
+     */
+    @ConfigItem(defaultValue = "true")
+    public boolean shared;
+
+    /**
+     * The value of the {@code quarkus-dev-service} label attached to the started container.
```
Comment (Ladicek) [inline extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaDevServicesBuildTimeConfig.java:51]:
```suggestion
     * The value of the {@code quarkus-dev-service-kafka} label attached to the started container.
```

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaDevServicesBuildTimeConfig.java [hunk 11]
```diff
@@ -33,4 +33,31 @@
...
+     * This property is used when {@code shared} is set to {@code true}.
+     * In this case, before starting a container, Dev Services for Kafka looks for a container with the
+     * {@code quarkus-dev-service}
```
Comment (Ladicek) [inline extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaDevServicesBuildTimeConfig.java:54]:
```suggestion
     * {@code quarkus-dev-service-kafka} label
```

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaDevServicesBuildTimeConfig.java [hunk 12]
```diff
@@ -33,4 +33,31 @@
...
+     * set to the configured value. If found, it will use this container instead of starting a new one. Otherwise it
+     * starts a new container with the {@code quarkus-dev-service} label set to the specified value.
```
Comment (Ladicek) [inline extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaDevServicesBuildTimeConfig.java:56]:
```suggestion
     * starts a new container with the {@code quarkus-dev-service-kafka} label set to the specified value.
```

**Comment by Ladicek** [comment]
Looks good, all my comments are about documentation (and one is about logging).

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java (hunk 8 above)
Comment (cescoffier) [reply extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java]: Fixed.

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaDevServicesBuildTimeConfig.java (hunk 9 above)
Comment (cescoffier) [reply extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaDevServicesBuildTimeConfig.java]: Fixed.

**Review by Ladicek — APPROVED** [review]

**Comment by Ladicek** [comment]
I know I'm a pain, but looks good now! :-)
